refactor: Add TypeScript support

fabiovincenzi opened the pull request:
**Overview**
This PR introduces TypeScript to git-proxy and refactors relevant code to support it.

**Changelog**
- Added TypeScript configuration (tsconfig.json) with essential settings for strict type checking, ES6 compatibility, and JSX support.
- Added typescript and ts-node to manage TypeScript code compilation and execution.
- Updated the package.json to include TypeScript dependencies.
- Converted the main entry file from JavaScript (index.js) to TypeScript (index.ts).
- Modified the start script to use ts-node for running TypeScript files.

Thanks to @jescalada for the work on fixing CI issues on https://github.com/G-Research/git-proxy/pull/31, which helped guide the necessary changes in this PR.

Related issue: #927

JamieSlome commented:
@06kellyjac - can we get a review and confirm whether this would constitute a patch or minor bump?

Any opine welcome @jescalada too 👍

06kellyjac commented on package.json:
```diff
@@ -6,9 +6,10 @@
     "client": "vite --config vite.config.js",
     "clientinstall": "npm install --prefix client",
-    "server": "node index.js",
+    "server": "tsx index.ts",
```
bin.git-proxy needs to be `index.ts`

06kellyjac commented on index.ts:
```diff
@@ -1,30 +1,37 @@
 #!/usr/bin/env node
```
```suggestion
#!/usr/bin/env tsx
```

node wont work anymore since this is typescript.

changing the things in the review then doing `npx .` works

06kellyjac replied:
could leave this header as node if we're going to build the ts to js. Or we could replace `tsx` with `node` in the resulting js

06kellyjac reviewed (changes requested):
These suggestions would ensure `npx @finos/git-proxy` continues to work.

Although the package in npm wouldn't work properly if someone is doing `require("@finos/git-proxy")` (or `import`) from plain JS. We probably want to build out valid JS with type defs for the published package

06kellyjac commented:
Squashing the commits down would be appreciated :slightly_smiling_face:

JamieSlome commented:
@fabiovincenzi - can you take a look at the comments and we can get this merged?

jescalada commented on index.ts:
@JamieSlome @06kellyjac I was wondering if we want to continue refactoring the project into ESM or just keep the CommonJS for now.

I've tried refactoring various modules to TS and ESM in subsequent PRs, but integration with Mocha and some ESM-only libraries (such as the `load-plugin` library in `/src/plugin.js`) causes many headaches and failing tests. I think converting into TS but keeping imports in CommonJS would be the least time-consuming.

On top of that, despite ESM being the "latest and greatest" way of doing things, there's a lot of resistance in the JS ecosystem because of integration problems with other libraries. Node 22 fixes these compatibility issues, but I'm not sure if it's acceptable to force git-proxy users to use Node 22.

**TL;DR:** Should we convert to ESM and bump Node to 22 (to fix compatibility issues in test runs), or keep CommonJS and only refactor modules into TS? 

I'd appreciate your thoughts on this! 😃

06kellyjac replied:
Thanks for the input. I do enjoy a good upgrade but since 20LTS has security support for 1 more year and 18LTS still has 1 month it'd probably be sensible to stick to commonjs for the time being.

If we're building out javascript from our typescript for the published NPM package I'm thinking sticking with commonjs shouldn't be much of a problem for us.

fabiovincenzi commented:
Hey @06kellyjac, @JamieSlome,

I've addressed all the comments and cleaned up the commit history by squashing the relevant commits.
Let me know if there's anything else needed to get this merged!

Thanks! 😊

06kellyjac commented:
No, I don't think this is importable from javascript as-is. We'll need to build out js and type definitions for the published package.

fabiovincenzi commented:
> No, I don't think this is importable from javascript as-is. We'll need to build out js and type definitions for the published package.

Hey @06kellyjac,

Could you clarify what you mean by "not importable from JavaScript as-is"? I thought git-proxy was being executed directly through the `index.js`, as the `package.json` doesn't have a "main" entry.

Also, do you want us to generate the JS and type definitions and include them in the publishing workflow for npm? Let me know how you'd like to proceed.

Thanks! 😊

fabiovincenzi commented:
Hi @06kellyjac, @JamieSlome,

I went ahead and implemented the changes to make the package importable from JavaScript/TypeScript:

1. **tsconfig.json**: enabled `"declaration": true` and set `outDir` to `"build/"`.  
2. **package.json**: added  
   - `"main": "build/index.js"`  
   - `"types": "build/index.d.ts"`  
3. **CI/build workflow**: now runs `npx tsc` to generate the `.js` and `.d.ts` files.  

Could you take a look and let me know if this covers the requirements you asked for?

Thanks! 😊

06kellyjac commented:
Thanks, I'll have a play with it.
One thing I'll need to test is importing other files. ATM we know some users are importing individual files to extend git-proxy so need to check that. Once everything is exported well via index we shouldn't have problems in that regard.

JamieSlome commented:
> Thanks, I'll have a play with it. One thing I'll need to test is importing other files. ATM we know some users are importing individual files to extend git-proxy so need to check that. Once everything is exported well via index we shouldn't have problems in that regard.

Thanks Jack. Keen to get this merged soon!

JamieSlome commented:
@jescalada - can we resolve conflicts? @06kellyjac - if no other issues, I'd like to merge this 👍

06kellyjac commented on package.json:
```diff
@@ -2,6 +2,8 @@
   "version": "1.11.0",
   "description": "Deploy custom push protections and policies on top of Git.",
+  "main": "dist/index.js",
+  "types": "dist/index.d.ts",
```
moving to `dist` would break existing usecases
not everything is exported or accessible via main/index atm

06kellyjac replied:
I'd love to go with this for a cut of v2 some time in the future

coopernetes commented:
I'm not sure what's going on with the diff on this branch. Possible replay of existing commits from main via merges? Can we get this branch rebased/cherry-picked with the specific TypeScript changes? I see a lot of unrelated changes that's making it difficult to fully review as [hinted by this comment](https://github.com/finos/git-proxy/pull/929#issuecomment-2690293570).
